### PR TITLE
Support advanced characters for entity name.

### PR DESCRIPTION
**Pull request by angelozerr**
Support advanced characters for entity name.

See https://github.com/redhat-developer/vscode-xml/issues/262

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
This PR fixes the issue  **Entity names with underscore and backspace** (the 4 issue from redhat-developer/vscode-xml#262).

I did some refactoring to use the same code for entity range diagnostics, completion, and definition

**Comment by fbricon** [comment]
I don't know if this is caused by this PR,  but given the following xml:

```xml
<!DOCTYPE web-app PUBLIC "-//Sun Microsystems, Inc.//DTD Web Application 2.3//EN" "http://java.sun.com/dtd/web-app_2_3.dtd"[
<!ENTITY mdash "&#x2014;">
]>
m| 
<web-app>
  <display-name>Servlet 2.3 Web Application</display-name>
</web-app>
```

 when typing anything outside the nodes, I get an NPE:

```
[Error - 2:10:18 PM] May 25, 2020 02:10:18 org.eclipse.lemminx.services.XMLCompletions collectOpenTagSuggestions()
Message: While performing ICompletionParticipant#onTagOpen
java.lang.IllegalArgumentException: Property must not be null: label
	at org.eclipse.lsp4j.util.Preconditions.checkNotNull(Preconditions.java:29)
	at org.eclipse.lsp4j.CompletionItem.<init>(CompletionItem.java:144)
	at org.eclipse.lemminx.extensions.contentmodel.participants.ContentModelCompletionParticipant.addCompletionItem(ContentModelCompletionParticipant.java:244)
	at org.eclipse.lemminx.extensions.contentmodel.participants.ContentModelCompletionParticipant.fillWithChildrenElementDeclaration(ContentModelCompletionParticipant.java:183)
	at org.eclipse.lemminx.extensions.contentmodel.participants.ContentModelCompletionParticipant.onTagOpen(ContentModelCompletionParticipant.java:65)
	at org.eclipse.lemminx.services.XMLCompletions.collectOpenTagSuggestions(XMLCompletions.java:608)
	at org.eclipse.lemminx.services.XMLCompletions.collectInsideContent(XMLCompletions.java:733)
	at org.eclipse.lemminx.services.XMLCompletions.doComplete(XMLCompletions.java:224)
	at org.eclipse.lemminx.services.XMLLanguageService.doComplete(XMLLanguageService.java:139)
	at org.eclipse.lemminx.XMLTextDocumentService.lambda$completion$1(XMLTextDocumentService.java:189)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1307)
	at java.base/java.util.concurrent.CompletableFuture$BiApply.tryFire(CompletableFuture.java:1276)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
```

There's no issue with vscode-xml 0.11.0

**Comment by fbricon** [comment]
Opened #720 about that last bug

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/entities/EntitiesCompletionExtensionsTest.java [hunk 1]
```diff
@@ -78,6 +78,24 @@ public void inside() throws BadLocationException {
 				c("&mdash;", "&mdash;", r(5, 2, 5, 14), "&mdash;"));
 	}
 
+	@Test
+	public void underscoreEntityName() throws BadLocationException {
+		// &foo_b|
+		String xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\r\n" + //
+				"<!DOCTYPE article [\r\n" + //
```
Comment (fbricon) [inline org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/entities/EntitiesCompletionExtensionsTest.java:85]: shouldn't it be `DOCTYPE  root` to be more consistent?

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/entities/EntitiesDefinitionExtensionsTest.java [hunk 2]
```diff
@@ -75,6 +75,19 @@ public void insideText() throws BadLocationException {
 		testDefinitionFor(xml, "test.xml", ll("test.xml", r(5, 7, 5, 12), r(2, 11, 2, 16)));
 	}
 
+	@Test
+	public void underscoreEntityName() throws BadLocationException {
+		String xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\r\n" + //
+				"<!DOCTYPE article [\r\n" + //
```
Comment (fbricon) [inline org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/entities/EntitiesDefinitionExtensionsTest.java:81]: shouldn't it be `DOCTYPE  root` to be more consistent?

Comment (angelozerr) [reply org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/entities/EntitiesDefinitionExtensionsTest.java]: fixed

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/entities/EntitiesCompletionExtensionsTest.java (hunk 1 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/entities/EntitiesCompletionExtensionsTest.java]: fixed

**Comment by angelozerr** [comment]
> I don't know if this is caused by this PR, but given the following xml:

Yes it's a a new bug that you have opened #720

**Comment by xorye** [comment]
When hovering over an entity that is not defined, I get documentation of the first defined entity:
image

xml:
```
<?xml version='1.0' encoding='UTF-8'?>
<!DOCTYPE article [
    <!ENTITY foo "foo">
]>

<article>
   &foo
</article>
```

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/XMLPositionUtility.java [hunk 3]
```diff
@@ -476,33 +482,74 @@ public static Range selectPreviousNodesEndTag(int offset, DOMDocument document)
...
+			return offset;
+		}
+		if (offset == 0) {
+			return -1;
+		}
+		int startEntityOffset = StringUtils.findStartWord(text, offset, ENTITY_NAME_PREDICATE);
+		if (startEntityOffset == -1) {
+			return -1;
+		}
+		if (startEntityOffset != -1 && text.charAt(startEntityOffset - 1) != '&') {
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/XMLPositionUtility.java:529]: Can `startEntityOffset == 0` here? In that case, I'm concerned about a possible IndexOutOfBoundsException here

**Comment by angelozerr** [comment]
> When hovering over an entity that is not defined, I get documentation of the first defined entity:

Sorry I push my current work for entities hover in this branch. I removed it, now you should not see hover.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/XMLPositionUtility.java (hunk 3 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/XMLPositionUtility.java]: Good catch! I have fixed the problem and write several tests with this method (see XMLPositionUtilityTest#entityReference()).

**Comment by xorye** [comment]
I found an edge case, which I think is caused by this code:
https://github.com/eclipse/lemminx/blob/f36d32a005400145905a8888d7a81d410b115db5/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/XMLPositionUtility.java#L551-L553

```
<?xml version='1.0' encoding='UTF-8'?>
<!DOCTYPE article [
    <!ENTITY hello "foo">
    <!ENTITY helloworld "foo">
]>

<article>
   &helloworld
```

When trying Go to definition in this case, with the cursor on the `o`, we go to the definition for `hello`

gif

I think this issue would be very rare, however

**Comment by angelozerr** [comment]
@xorye I change the strategy to discover entity. The referenced entity must be ends with ';'. So in your case definition will not work. You must write ';' at the end of helloword.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/DTDErrorCode.java [hunk 4]
```diff
@@ -36,41 +37,17 @@
...
-	NotationDeclUnterminated,
-	OpenQuoteExpected,
-	OpenQuoteMissingInDecl,
-	PEReferenceWithinMarkup,
-	QuoteRequiredInPublicID,
-	QuoteRequiredInSystemID,
-	SpaceRequiredAfterSYSTEM,
-	dtd_not_found("dtd-not-found");
-			
+	AttNameRequiredInAttDef, AttTypeRequiredInAttDef, ElementDeclUnterminated, EntityDeclUnterminated,
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/DTDErrorCode.java:40]: please fix formatting

**Comment by fbricon** [comment]
> @xorye I change the strategy to discover entity. The referenced entity must be ends with ';'. So in your case definition will not work. You must write ';' at the end of helloword.

you could also detect if there's a whitespace

**Comment by angelozerr** [comment]
> you could also detect if there's a whitespace

I think it's a bad idea, because entity reference must start with '&' and ends with ';'

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/DTDErrorCode.java (hunk 4 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/DTDErrorCode.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/XMLPositionUtility.java [hunk 5]
```diff
@@ -476,33 +503,103 @@ public static Range selectPreviousNodesEndTag(int offset, DOMDocument document)
...
+		return selectEntityReference(offset, document, true);
+	}
+
+	/**
+	 * Returns the range of the entity reference in a text node (ex : &amp;) and
+	 * null otherwise.
+	 * 
+	 * @param offset            the offset
+	 * @param document          the document
+	 * @param endsWithSemicolon true if the entity reference must be ends with ';'
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/XMLPositionUtility.java:524]: ... must end with ...

Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/XMLPositionUtility.java]: fixed
